### PR TITLE
Fix Automatic Proxy Configuration protocol for OSX 10.13

### PR DESCRIPTION
**Pull request by iamphi**
Reference comparison of kCFProxyTypeAutoConfigurationURL doesn't work in OSX 10.13. Added fallback to do value/string comparison to get proxy type.

In OSX 10.13, the IntPrt value for kCFProxyTypeKey in the dictionary returned by CFProxyAutoConfigurationResultCallback no longer match CFProxy.kCFProxyTypeAutoConfigurationURL.

I tested for my use case by compiling the mono 5.8.0.127 tarball with my changes.

Excerpt of my discovery code (altered code from CFNetwork.ExecuteProxyAutoConfigurationURL) :
```CSharp
        [DllImport(CFObject.CoreFoundationLibrary)]
        extern static void CFDictionaryGetKeysAndValues(IntPtr theDict, IntPtr[] keys, IntPtr[] values);

        [DllImport(CFObject.CoreFoundationLibrary)]
        extern static int CFDictionaryGetCount(IntPtr theDict);

...

           var proxyAutoConfigURL = CFUrl.Create("http://[url-to-doc-pac-file]");
            var targetURL = CFUrl.Create("http://[target-url]");

            CFProxy[] proxies = null;

            var runLoop = CFRunLoop.CurrentRunLoop;

            // Callback that will be called after executing the configuration script
            CFNetwork.CFProxyAutoConfigurationResultCallback cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error)
            {
                Console.WriteLine("In Callback");
                if (proxyList != IntPtr.Zero)
                {
                    var array = new CFArray(proxyList, false);
                    proxies = new CFProxy[array.Count];
                    for (int i = 0; i < proxies.Length; i++)
                    {
                        CFDictionary dict = new CFDictionary(array[i], false);
                        var v = dict.GetValue(CFProxy.kCFProxyTypeKey);
                        Console.WriteLine($"kCFProxyTypeKey {CFString.AsString(v)}");
                        var count = CFDictionaryGetCount(dict.Handle);

                        var keys = new IntPtr[count];
                        var values = new IntPtr[count];
                        CFDictionaryGetKeysAndValues(dict.Handle, keys, values);

                        foreach(var k in keys) {
                            var keyStr = new CFString(k, false);
                            Console.WriteLine($"{keyStr}");
                        }

                        proxies[i] = new CFProxy(dict);
                    }
                    array.Dispose();
                }
                runLoop.Stop();
            };

            var clientContext = new CFStreamClientContext();
            var loopSource = CFNetwork.CFNetworkExecuteProxyAutoConfigurationURL(proxyAutoConfigURL.Handle, targetURL.Handle, cb, ref clientContext);

            // Create a private mode
            var mode = CFString.Create("Mono.MacProxy");

            runLoop.AddSource(loopSource, mode);
            runLoop.RunInMode(mode, double.MaxValue, false);
            runLoop.RemoveSource(loopSource, mode);

            foreach (var p in proxies)
            {
                Console.WriteLine($"{p.ProxyType} {p.HostName}");
            }
```

### CODE REVIEW

**Comment by marek-safar** [comment]
approve

**Comment by marek-safar** [comment]
@chamons could you chime in?

File: mcs/class/System/System.Net/MacProxy.cs [hunk 1]
```diff
@@ -688,6 +688,24 @@ static CFProxyType CFProxyTypeToEnum (IntPtr type)
 			if (type == kCFProxyTypeSOCKS)
 				return CFProxyType.SOCKS;
 			
+			//in OSX 10.13 pointer comparison didn't work for kCFProxyTypeAutoConfigurationURL
+			var typeString = CFString.AsString(type);
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Is this a cast or an allocate? Do we need to clean it up?

**Comment by chamons** [comment]
Apple sometimes using pointer comparisons and sometimes using string comparisons against constants is not unheard of. Normally it's the other way around, but this seems reasonable.

I have one question on the memory allocation (or lack of) on CFString.AsString, but beyond that lgtm.

File: mcs/class/System/System.Net/MacProxy.cs (hunk 1 above)
Comment (iamphi) [reply mcs/class/System/System.Net/MacProxy.cs]: I'll change it use CFStringCompare

https://developer.apple.com/documentation/corefoundation/1542911-cfstringcompare?language=objc

File: mcs/class/System/System.Net/MacProxy.cs [hunk 2]
```diff
@@ -335,6 +335,15 @@ public static CFString Create (string value)
 				return (int) CFStringGetLength (Handle);
 			}
 		}
+		
+		[DllImport (CoreFoundationLibrary)]
+		//CFComparisonResult CFStringCompare(CFStringRef theString1, CFStringRef theString2, CFStringCompareFlags compareOptions)
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Remove comment

File: mcs/class/System/System.Net/MacProxy.cs [hunk 3]
```diff
@@ -335,6 +335,15 @@ public static CFString Create (string value)
...
+		extern static int CFStringCompare(IntPtr theString1, IntPtr theString2, int compareOptions);
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: space before (

File: mcs/class/System/System.Net/MacProxy.cs [hunk 4]
```diff
@@ -335,6 +335,15 @@ public static CFString Create (string value)
...
+		
+		public static int Compare(IntPtr string1, IntPtr string2, int compareOptions = 0)
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Space before (

File: mcs/class/System/System.Net/MacProxy.cs [hunk 5]
```diff
@@ -335,6 +335,15 @@ public static CFString Create (string value)
...
+		{
+			return CFStringCompare(string1, string2, compareOptions);
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Same

File: mcs/class/System/System.Net/MacProxy.cs [hunk 6]
```diff
@@ -689,22 +698,23 @@ static CFProxyType CFProxyTypeToEnum (IntPtr type)
...
-				case "kCFProxyTypeFTP":         
-					return CFProxyType.FTP;     
-				case "kCFProxyTypeHTTP":        
-					return CFProxyType.HTTP;    
-				case "kCFProxyTypeHTTPS":       
-					return CFProxyType.HTTPS;   
-				case "kCFProxyTypeSOCKS":       
-					return CFProxyType.SOCKS;   
-			}
+			if (CFString.Compare(type, kCFProxyTypeAutoConfigurationJavaScript) == 0)
```
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Same through these.

**Pull request merged**